R/utility_spliceHelper_functions.R

```diff
@@ -20,220 +20,336 @@
+#' used to enable the comparison of ranges with all elements in a
+#' rangesList, for example to compute the overlap of first and last
+#' exons with matching transcripts (see compareTranscripts()).
+#' @param ranges an IRanges object
+#' @param target an IRangesList object
+#' @return a ranges object with mcols objects
+#' \itemize{
+#'    \item matchRng repeated matched ranges
+#'    \item IdMap index of the repeated matched ranges
+#' }
+#' @noRd
+expandRanges <- function(ranges,target){ 
+  processedRng <- rep(ranges,elementNROWS(target))
+  mcols(processedRng)$IdMap <- rep(1:length(ranges),elementNROWS(target))
+  mcols(processedRng)$matchRng <- unlist(target, use.names=F)
+  return (processedRng)
+}
+
+#' splice ranges pre-processing
+#' @description this function takes in an IRangesList object and a 
+#' target IRangesList object with the same length, where each list
+#' entry in target IRangesList[[i]] (target[[i]]) corresponds to the
+#' matching list entry in IRangesList[[i]] (rglist[i]). The function
+#' then creates an IRanges object with a length corresponding to the
+#' length(unlist(target))*length(unlist(rglist)), where each individual
+#' element in rglist[[i]] is repeated to match each individual element
+#' in target[[i]]. The repeated ranges are stored in mcols()$matchRng 
+#' while the corresponding index i for each element of 
+#' target[[i]]*rglist[[i]] is stored in mcols()$IdMap. This function is
+#' used to enable the comparison of ranges with all elements in a
+#' rangesList, for example to compute the overlap of splice sites
+#' with matching transcripts (see compareTranscripts()).
+#' @param rglist an IRangesList object
+#' @param target an IRangesList object
+#' @return a ranges object with mcols objects
+#' \itemize{
+#'    \item matchRng repeated matched ranges
+#'    \item IdMap index of the repeated matched ranges
+#' }
+#' @noRd
+expandRangesList <- function(rglist,target){ 
+  processedRng <- rep(unlist(rglist, use.names=F),rep(elementNROWS(target),times=elementNROWS(rglist)))
+  mcols(processedRng)$IdMap <- rep(1:length(rglist),elementNROWS(rglist)*elementNROWS(target))
+  mcols(processedRng)$matchRng <- unlist(rep(target, times= elementNROWS(rglist)), use.names=F)
+  return (processedRng)
+}
+
+#' alternative start/end exon
+#' @description This function checks whether an 
+#' alternative start/end exon is used by overlapping
+#' the exon ranges of the first (or last) exons of matching transcripts.
+#' @noRd
+alternativeStartEndExon <- function(queryRng, subjectRng){
+  return(!poverlaps(queryRng, subjectRng))
+}
+
+#' alternative TSS distance
+#' @description This function calculates the distance
+#' of an alternative TSS by comparing the start
+#' coordinates of the start exon ranges of matching
+#' transcripts. If an alternative first exon is used the distance is set to 0.
+#' @noRd
+calculateTSSdistance <- function(queryStartRng, subjectStartRng,
```

Comment:
Hi @yuukiiwa, I am not sure if this is a good or bad idea, I feel that calculateTSSdistance and calculateTESdistance can potentially be collapsed into one function, as they are highly similar except the internal function used is slightly different(one is using start, the other is using end). If I were you, I probably will create a function call `calculateTerminalDistance <- function(queryTerminalExonRng, ...alternativeTerminalExonm, strand, direction = "Start")`, and use the direction to determine which function to be used

Reply:
With this direction variable, we then need an if...else... to separate the operations for TSS and TES, which will not be as time efficient.

Reply:
You are right maybe! If so, I guess you can leave this function as it is as well!

Reply:
Sorry for coming back to this again, I realized that there is some advanced way that you can do to avoid the `if` if you feel it's less time efficient?:
direction_names <- c("Start","End")
if direction = "Start", the first line can be like this 
`.... <- get(tolower(direction))()` 
the second line will be 
"....[strand == "-"]  <- get(tolower(setdiff(direction_names, direction)))()" 

'get' is a function to use for variable function names

Reply:
I have merged the two functions.